lib/ddtrace/contrib/rest_client/request_patch.rb

```diff
@@ -62,11 +62,11 @@ def datadog_trace_request(uri)
             # rubocop:disable Lint/RescueException
           rescue Exception => e
             # rubocop:enable Lint/RescueException
-            span.set_error(e)
+            span.set_error(e) if span
 
             raise e
           ensure
-            span.finish
+            span.finish if span
```

Comment:
we call
```
ensure
  span.finish
```
and, i believe, `span.set_error(e)` , in a number of instrumentation `rescue` / `ensure` blocks, such as action_controller: https://github.com/DataDog/dd-trace-rb/blob/9976f36da4c8eabb8dc3d370e1c48f6cd89b9f5d/lib/ddtrace/contrib/action_pack/action_controller/instrumentation.rb#L67-L72

 Is there something that makes `rest_client` unique here where we have to be extra defensive about span being nil, which is handled in other instrumentations?

Reply:
Yes, the main difference is that we have a catch-all `rescue Exception => e` block here, which can swallow non-StandardError exceptions and cause `ensure; span.finish` to throw a StandardError (like `NoMethodError: undefined method `finish' for nil:NilClass`) instead of properly propagating the non-StandardError up the stack. 

This could cause the application to hang during graceful shutdown, or just completely misbehave in case of an out-of-resource error (e.g. `SystemStackError`, `NoMemoryError`).

This doesn't mean cases like the one you linked are ok, but they are won't cause as much of critical issues as the ones addressed here.

Reply:
gotcha ty for clarification